// from-wechat/tool.js
var tool = {
    isLog: !1,
    log: function() {
        this.isLog && console.log(arguments)
    },
    now: function() {
        return +new Date
    },
    getCookie: function(e) {
        for (var t = e + "=", o = document.cookie.split(";"), n = 0; n < o.length; n++) {
            for (var r = o[n]; " " == r.charAt(0); )
            r = r.substring(1);
            if (-1 != r.indexOf(t))
                return r.substring(t.length, r.length)
        }
        return ""
    },
    setCookie: function(e, t, o) {
        var n = new Date;
        n.setTime(n.getTime() + 24 * o * 60 * 60 * 1e3);
        var r = "expires=" + n.toUTCString();
        document.cookie = e + "=" + t + "; " + r
    },
    clearCookie: function() {
        for (var e = document.cookie.split(";"), t = 0; t < e.length; t++) {
            var o = e[t]
                , n = o.indexOf("=")
                , r = n > -1 ? o.substr(0, n) : o;
            document.cookie = r + "=;expires=Thu, 01 Jan 1970 00:00:00 GMT"
        }
    },
    getLocalStorage: function() {
        return window.localStorage || {
            getItem: function() {
                return void 0
            },
            setItem: function() {},
            removeItem: function() {},
            key: function() {
                return ""
            }
        }
    },
    htmlEncode: function(e) {
        return angular.isString(e) ? e.replace(/&/g, "&amp;").replace(/"/g, "&quot;").replace(/'/g, "&#39;").replace(/</g, "&lt;").replace(/>/g, "&gt;") : ""
    },
    htmlDecode: function(e) {
        return e && 0 != e.length ? e.replace(/&lt;/g, "<").replace(/&gt;/g, ">").replace(/&#39;/g, "'").replace(/&quot;/g, '"').replace(/&amp;/g, "&") : ""
    },
    hrefEncode: function(e) {
        var t = this
            , o = e.match(/&lt;a href=(?:'|").*?(?:'|").*?&gt;.*?&lt;\/a&gt;/g);
        if (o) {
            for (var n, r, a = 0, c = o.length; c > a; ++a)
            n = /&lt;a href=(?:'|")(.*?)(?:'|").*?&gt;.*?&lt;\/a&gt;/.exec(o[a]),
            n && n[1] && (r = n[1],
                          t.isUrl(r) && (e = e.replace(n[0], this.htmlDecode(n[0])).replace(n[1], u.genCheckURL(n[1]))));
                          return e
        }
        return e.replace(new RegExp(i,"ig"), function() {
            return '<a target="_blank" href="' + u.genCheckURL(arguments[0].replace(/^(\s|\n)/, "")) + '">' + arguments[0] + "</a>"
        })
    },
    clearHtmlStr: function(e) {
        return e ? e.replace(/<[^>]*>/g, "") : e
    },
    clearLinkTag: function(e) {
        return e
    },
    setCheckUrl: function(e) {
        r = "&skey=" + encodeURIComponent(e.getSkey()) + "&deviceid=" + encodeURIComponent(e.getDeviceID()) + "&pass_ticket=" + encodeURIComponent(e.getPassticket()) + "&opcode=2&scene=1&username=" + e.getUserName()
    },
    genCheckURL: function(e) {
        if (!r)
            throw "_checkURLsuffix is not ready!";
        return o.API_webwxcheckurl + "?requrl=" + encodeURIComponent((0 == e.indexOf("http") ? "" : "http://") + u.clearHtmlStr(u.htmlDecode(e))) + r
    },
    isUrl: function(e) {
        return new RegExp(i,"i").test(e)
    },
    formatNum: function(e, t) {
        var o = (isNaN(e) ? 0 : e).toString()
            , n = t - o.length;
        return n > 0 ? [new Array(n + 1).join("0"), o].join("") : o
    },
    getServerTime: function() {
        return (new Date).getTime()
    },
    globalEval: function(e) {
        e && /\S/.test(e) && (window.execScript || function(e) {
            window.eval.call(window, e)
        }
                             )(e)
    },
    evalVal: function(e) {
        var t, o = "a" + this.now();
        return this.globalEval(["(function(){try{window.", o, "=", e, ";}catch(_oError){}})();"].join("")),
        t = window[o],
        window[o] = null ,
        t
    },
    browser: function() {
        var e, t = navigator.userAgent.toLowerCase();
        if (null  != t.match(/trident/))
            e = {
                browser: "msie",
                version: null  != t.match(/msie ([\d.]+)/) ? t.match(/msie ([\d.]+)/)[1] : t.match(/rv:([\d.]+)/)[1]
            };
            else {
                var o = /(msie) ([\w.]+)/.exec(t) || /(chrome)[ \/]([\w.]+)/.exec(t) || /(webkit)[ \/]([\w.]+)/.exec(t) || /(opera)(?:.*version|)[ \/]([\w.]+)/.exec(t) || t.indexOf("compatible") < 0 && /(mozilla)(?:.*? rv:([\w.]+)|)/.exec(t) || [];
                e = {
                    browser: o[1] || "",
                    version: o[2] || "0"
                }
            }
            var n = {};
            return e.browser && (n[e.browser] = !0,
                                 n.version = e.version),
                                 n.chrome ? n.webkit = !0 : n.webkit && (n.safari = !0),
                                 n
    }(),
    isSpUser: function(e) {
        for (var t = 0, o = c.length; o > t; t++)
        if (c[t] === e || /@qqim$/.test(e))
            return !0;
        return !1
    },
    isShieldUser: function(e) {
        if (/@lbsroom$/.test(e) || /@talkroom$/.test(e))
            return !0;
        for (var t = 0, o = s.length; o > t; ++t)
        if (s[t] == e)
            return !0;
        return !1
    },
    isRoomContact: function(e) {
        return e ? /^@@|@chatroom$/.test(e) : !1
    },
    getContactHeadImgUrl: function(e) {
        return (this.isRoomContact(e.UserName) ? o.API_webwxgetheadimg : o.API_webwxgeticon) + "?seq=0&username=" + e.UserName + "&skey=" + e.Skey + (e.MsgId ? "&msgid=" + e.MsgId : "") + (e.EncryChatRoomId ? "&chatroomid=" + e.EncryChatRoomId : "")
    },
    form: function(e, t) {
        t = t || {};
        var o, n = [];
        n.push('<form method="POST" action="' + this.htmlEncode(e) + '">');
        for (var r in t)
            n.push('<input type="hidden" name="' + r + '" value="' + t[r] + '">');
        n.push("</form>"),
        o = angular.element(n.join(""))[0],
        document.body.appendChild(o),
        o.submit()
    },
    queryParser: function() {
        for (var e = {}, t = location.search.substring(1), o = t.split("&"), n = 0, r = o.length; r > n; n++) {
            var a = o[n].split("=")
                , i = decodeURIComponent(a[0]);
            e[i] = decodeURIComponent(a[1] || "")
        }
        return e
    },
    getSize: function(e) {
        if (e = +e) {
            var t = 10
                , o = 10
                , n = 20
                , r = 1 << o
                , a = 1 << n;
            if (e >> n > 0) {
                var i = Math.round(e * t / a) / t;
                return "" + i + "MB"
            }
            if (e >> o - 1 > 0) {
                var c = Math.round(e * t / r) / t;
                return "" + c + "KB"
            }
            return "" + e + "B"
        }
    },
    xml2json: function(e) {
        if (!e)
            return {};
        try {
            var t = e.indexOf("<");
            return t && (e = e.substr(t)),
            $.xml2json(e)
        } catch (o) {
            return console.error(o),
            {}
        }
    },
    encodeEmoji: function(e) {
        return e = e || "",
        e = e.replace(/<span class="(emoji emoji[a-zA-Z0-9]+)"><\/span>/g, "###__EMOJI__$1__###")
    },
    decodeEmoji: function(e) {
        return e = e || "",
        e = e.replace(/###__EMOJI__(emoji emoji[a-zA-Z0-9]+)__###/g, '<span class="$1"></span>')
    },
    removeHtmlStrTag: function(e) {
        return e = e || "",
        e = this.encodeEmoji(e),
        e = this.htmlDecode(e),
        e = this.clearHtmlStr(e),
        e = this.decodeEmoji(e)
    },
    safeDigest: function(e) {
        e = e || t,
        e.$$phase || e.$digest()
    },
    wait: function(e, t, o) {
        var o = o || 10;
        setTimeout(function n() {
            e() ? t() : setTimeout(n, o)
        }, o)
    },
    // fitRun: n,
    findIndex: function(e, t) {
        for (var o = 0; o < e.length; o++)
        if (e[o] == t)
            return o;
        return -1
    },
    genEmoticonHTML: function(e, t) {
        return '<img class="' + e + '" text="' + t + (t.indexOf(o.MM_EMOTICON_WEB) > -1 ? "" : o.MM_EMOTICON_WEB) + '" src="' + o.RES_IMG_PLACEHOLDER + '" />'
    },
    getShareObject: function(e) {
        return a[e] = a[e] || {},
        a[e]
    },
    isUserName: function() {},
    isWindows: /windows/gi.test(navigator.userAgent),
    isMacOS: /macintosh/gi.test(navigator.userAgent),
    isIPad: /ipad/gi.test(navigator.userAgent)
};

export default tool
